### PR TITLE
[DependencyInjection] Make extensions alias case-insensitive

### PR DESCRIPTION
**Pull request by dosten**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | yes |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14212 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/ProjectWithUpperAliasAndNamespaceExtension.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+use Symfony\Component\DependencyInjection\Extension\Extension;
+
+class ProjectWithUpperAliasAndNamespaceExtension extends Extension
+{
+    public function load(array $configs, ContainerBuilder $configuration) {}
+
+    public function getNamespace()
+    {
+        return 'http://www.example.com/schema/PROJECT';
+    }
```
Comment (WouterJ) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/ProjectWithUpperAliasAndNamespaceExtension.php]: the namespace isn't important for this, is it? (and namespace is not used for internal storage by the container).

**Comment by WouterJ** [comment]
You should document this change in the CHANGELOG file of the DependencyInjection component. Something like: `* [BC BREAK] The extension name and namespace are now case-insentive.`

**Comment by dosten** [comment]
@WouterJ I added the entry into the CHANGELOG.md file

**Comment by dosten** [comment]
@WouterJ `ContainerBuilder::getExtension()` can search by name or namespace, so i think it's logical to lowercase the namespace to keep consistency.

**Comment by hacfi** [comment]
I like it :)

**Comment by Tobion** [comment]
XML namespace are case-sensitive.

>  IRI references identifying namespaces are compared when determining whether a name belongs to a given namespace, and whether two names belong to the same namespace. [Definition: The two IRIs are treated as strings, and they are identical if and only if the strings are identical, that is, if they are the same sequence of characters. ] The comparison is case-sensitive, and no %-escaping is done or undone.

http://www.w3.org/TR/xml-names11/

**Comment by dosten** [comment]
@Tobion updated the PR to only make the extension alias case-insensitive.

**Comment by Tobion** [comment]
the phpdoc should be updated to reflect the changes

**Comment by dosten** [comment]
@Tobion done

**Comment by Tobion** [comment]
:+1:

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 2]
```diff
@@ -152,8 +153,8 @@ public function registerExtension(ExtensionInterface $extension)
      */
     public function getExtension($name)
     {
-        if (isset($this->extensions[$name])) {
-            return $this->extensions[$name];
+        if (isset($this->extensions[strtolower($name)])) {
```
Comment (stloyd) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]:
```php
$name = strtolower($name);
if (isset($this->extensions[$name])) {
    return $this->extensions[$name];
}

Comment (dosten) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: If i do that the search by namespace will be case-insensitive

Comment (inso) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]:
```PHP
$lowercaseName = strtolower($name);
if (isset($this->extensions[$lowercaseName])) {
    return $this->extensions[$lowercaseName];
}
```

Comment (dosten) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: @inso done!

**Comment by dosten** [comment]
Rebased and squashed. The failing tests aren't related.

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+3.0.0
+-----
+
+ * [BC BREAK] The extension alias is now case-insentive.
```
Comment (jrobeson) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: typo .. insensitive.

**Comment by Tobion** [comment]
Considering there have been recent PRs to lower the number on strtolower calls for the container, which led to complicated and ugly code, how about making the container case-sensitive in symfony 3.0?
I don't really thing case-insensitivity is needed. Also it is not consistent with other places where we have names, e.g. the route names are also not case-insensitive!

**Comment by jakzal** [comment]
> how about making the container case-sensitive in symfony 3.0

Currently there's a method generated for each service based on its id. Method names are case insensitive in PHP.  How would you implement case sensitivity without introducing a big revolution in the component?

**Comment by GromNaN** [comment]
@jakzal The component could be strict with id case, but forbid usage of the same service id with different cases.

**Comment by jakzal** [comment]
@GromNaN I thought about it, but that means we still need to do strtolower.

**Comment by dosten** [comment]
Any comment? The fact that method names are case insensitive would be a big drawback if we want to make the container case sensitive. IMO we should keep it case insensitive.

**Comment by dosten** [comment]
Any feedback about this?

**Comment by xabbuh** [comment]
> Currently there's a method generated for each service based on its id. Method names are case insensitive in PHP. How would you implement case sensitivity without introducing a big revolution in the component?

But the user will never call these methods directly, will they? So we can handle services case-sensitive by controlling how the method map is accessed in the container's `get()` method.

**Comment by dosten** [comment]
So, what's the veredict? case-sensitive or case-insensitive?

**Comment by stof** [comment]
> But the user will never call these methods directly, will they?

This is not about calling these methods by the user, but about avoiding conflicts between several service ids generating the same method name.
What we would have to do is forbidding to have 2 services ids registered in the container and different only by their case. Doing this would still require as much overhead as today in methods building the container, and maybe even more as we would have to keep the user-entered case in the list of services, making the logic to check case-conflicting services more complex than when all ids are registered in lowercase (we would have to store a second array mapping lowercase ids to the exact case being used)

**Comment by stof** [comment]
however, I'm not sure making the DI alias case sensitive is actually worth it.

**Pull request closed without merging**